### ISSUE TITLE
top/bottom materializations should take bounds/constraints into account

### ISSUE THREAD

**Issue by adamjstewart** [astral-sh/ty#2660]
### Summary

Assert narrowing works when only one type is checked, but not when a union of types is checked. Example code:
```python
import os

def foo(src: str | os.PathLike | list[str | os.PathLike]) -> None:
    assert isinstance(src, str | os.PathLike)
    os.makedirs(src)
```
This fails with:
```console
> ty check test.py
error[invalid-argument-type]: Argument to function `makedirs` is incorrect
 --> test.py:5:17
  |
3 | def foo(src: str | os.PathLike | list[str | os.PathLike]) -> None:
4 |     assert isinstance(src, str | os.PathLike)
5 |     os.makedirs(src)
  |                 ^^^ Expected `str | bytes | PathLike[str] | PathLike[bytes]`, found `str | PathLike[Unknown] | (list[str | PathLike[Unknown]] & PathLike[object])`
  |
info: Element `list[str | PathLike[Unknown]] & PathLike[object]` of this union is not assignable to `str | bytes | PathLike[str] | PathLike[bytes]`
info: Function defined here
    --> stdlib/os/__init__.pyi:1906:5
     |
1904 |         """
1905 |
1906 | def makedirs(name: StrOrBytesPath, mode: int = 0o777, exist_ok: bool = False) -> None:
     |     ^^^^^^^^ -------------------- Parameter declared here
1907 |     """makedirs(name [, mode=0o777][, exist_ok=False])
     |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```
Playground link: https://play.ty.dev/e641e063-a2de-4af7-9566-046ce007f2bd

Apologies if this is a duplicate of an existing issue, I did look but didn't see anything identical.

### Version

ty 0.0.14

**Comment by carljm** [comment]
We do support narrowing on calls to `isinstance` with unions of types, as shown in [this playground](https://play.ty.dev/6eef7b85-446e-4f87-885a-949a539bbbda).

What's happening here is a bit more subtle. Narrowing is happening, but the narrowed type includes `(list[str | PathLike[Unknown]] & PathLike[object])` -- the intersection of a list and an `os.PathLike` (e.g. a class that inherits both `list` and `os.PathLike`).

But I think there is a bug in ty here. The call to `makedirs` fails because this intersection is with `PathLike[object]`. This is ty saying "when you check that something is an instance of PathLike, I still have no idea what _kind_ of PathLike it is, so I'll use `PathLike[object]` which covers all the possibilities." But what ty is missing here is that PathLike is generic over a constrained typevar, which can only represent `str` or `bytes`. Thus `PathLike[object]` is too wide a type -- I think we should really use `PathLike[str] | PathLike[bytes]` here instead, because those are the only actual possibilities. And if we did that, the `makedirs` call would succeed, since it accepts both `PathLike[str]` and `PathLike[bytes]`.

Putting this in more general and theoretical terms: given a generic class `C[T]`, today the top and bottom materializations of `C[Unknown]` always look like this:

| | Covariant | Invariant | Contravariant |
|--|--|--|--|
| **Top** | `C[object]` | `Top[C[Unknown]]` | `C[Never]` |
| **Bottom** | `C[Never]` | `Bottom[C[Unknown]]` | `C[object]` |

But this should take upper bounds and constraints into account! If `T` has an upper bound `B` (`C[T: B]`), then the chart should look like this instead:

| | Covariant | Invariant | Contravariant |
|--|--|--|--|
| **Top** | `C[B]` | `Top[C[Unknown]]` | `C[Never]` |
| **Bottom** | `C[Never]` | `Bottom[C[Unknown]]` | `C[B]` |

And (this is the part that would fix this example), if `T` has constraints (e.g. `C[T: (str, bytes)]`), then it should look like this:

| | Covariant | Invariant | Contravariant |
|--|--|--|--|
| **Top** | `C[str] \| C[bytes]` | `Top[C[Unknown]]` | `C[Never]` |
| **Bottom** | `C[Never]` | `Bottom[C[Unknown]]` | `C[str] \| C[bytes]` |

In both cases, the `Never` version doesn't change because we don't have a way to express that type, so we stick with the less-precise version; it's less likely to matter.

**Comment by AlexWaygood** [comment]
This is similar to https://github.com/astral-sh/ty/issues/1164, which is about our top materializations looking at type variable defaults when they should instead be looking at the upper bounds of type variables

**Comment by carljm** [comment]
And it's a duplicate of #1109
